File: 2020/ruby/23/solve.rb
require 'set'

if !true
  a = File.readlines('sampleinput.txt').map(&:chomp)
else
  a = File.readlines('input.txt').map(&:chomp)
end

if !true
  def p(*args)
    $a = args
  end
  at_exit { Kernel.p(*$a) }
end

# a = a.slice_when { |f| f == '' }.to_a.map { |f|
#   if f.last == ''
#     f.pop
#   end
#   f
# }

cups = a.first.split('').map(&:to_i)

100.times do

  cc = cups.first
  cs = cups.drop(1).take(3)
  cups = cups.take(1) + cups.drop(4)

  finder = cc - 1
  loop do
    if finder < cups.min
      finder = cups.max
    end

    dest = cups.find { |c| c == finder }
    if dest
      break
    else
      finder -= 1
    end
  end

  p "DEST", finder
  fi = cups.index(finder)

  cups = cups.take(fi + 1) + cs + cups.drop(fi + 1)
  cups = cups.drop(1) + cups.take(1)
end

fi = cups.index(1)

cups = cups.drop(fi) + cups.take(fi)

p cups.drop(1).join

File: 2020/ruby/23/solve.scala
var cups = new scala.collection.mutable.DoubleLinkedList[Integer]
cups.append(Source.fromFile("sampleinput.txt").getLines().head.split("").map(_.toInt).toList)

val CUPS = 999999
val TIMES = 10 * 1000 * 1000

cups += (cups.max).upto(CUPS)

(1 to TIMES).foreach {
  val cc = cups.head
  cups.remove()
  val newCups = cups.drop(1).take(3)

  val finder = if (cc - 1 == 0) CUPS else cc - 1
  val cursor = cups.dropWhile(_ != finder)
  cursor.insert(newCups)

  cups.append(scala.collection.mutable.LinkedList[Integer](cc))
}

print(cups)
